### ISSUE TITLE
configure_file "looses" it's dependencies

### ISSUE THREAD

**Issue by gerion0** [mesonbuild/meson#12826]
**Describe the bug**
When using `configure_file` with `configuration_data` objects, it is possible to include paths to other targets.
However, there seems no way to also declare a dependency between `configure_file` and the other target then.

**To Reproduce**
Here is a minimal example (it needs three files):

`test.sh.in`
```sh
cat @test_file@
```

`testfile`
```
foo
```

`meson.build`
```meson
project('test', 'c')

fs = import('fs')

some_tgt = fs.copyfile(files('testfile')[0], 'testfile')
conf_data = configuration_data()
conf_data.set('test_file', some_tgt.full_path())

sh = configure_file(input: files('test.sh.in'),
                    output: 'test.sh',
                    configuration: conf_data)

run_target('cat', command: [find_program('bash'), sh])
```

The `run_target` fails then with the following error:
```
$ meson setup build
The Meson build system
Version: 1.3.1
Source dir: /home/gerion/test/mesonbug
Build dir: /home/gerion/test/mesonbug/build
Build type: native build
Project name: test
Project version: undefined
C compiler for the host machine: ccache cc (gcc 13.2.1 "cc (Gentoo Hardened 13.2.1_p20240113-r1 p12) 13.2.1 20240113")
C linker for the host machine: cc ld.bfd 2.40
Host machine cpu family: x86_64
Host machine cpu: x86_64
Configuring test.sh using configuration
Program bash found: YES (/usr/bin/bash)
Build targets in project: 2

Found ninja-1.11.1 at /usr/bin/ninja
$ cd build; meson compile cat
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja cat
[0/1] Running external command cat (wrapped by meson to set env)
cat: /home/gerion/test/mesonbug/build/testfile: No such file or directory
FAILED: meson-internal__cat 
env MESON_SOURCE_ROOT=/home/gerion/test/mesonbug MESON_BUILD_ROOT=/home/gerion/test/mesonbug/build MESON_SUBDIR= 'MESONINTROSPECT=/usr/lib/python-exec/python3.12/meson introspect' /usr/bin/bash test.sh
ninja: build stopped: subcommand failed.
```

**Expected behavior**
Either Meson should set the dependency between the `configuration_data`/`configure_file` and `some_tgt` implicitely (what is the meaning of taking a `full_path()` of a non existent file?) or give the user the opportunity to specify an explicit dependency as part of `conf_data.set()` or `configure_file()`.

The above example can be "fixed" by specifying `depends: some_tgt` within the `run_target`. However, this is not the direct way and could be tedious. For example, one could use `sh` in *a lot of* targets and would need to specify the explicit dependency then in every of these targets.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?: Native
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.): Gentoo Linux
* what Python version are you using e.g. 3.8.0: 3.12
* what `meson --version`: 1.3.1
* what `ninja --version` if it's a Ninja build: 1.11.1

**Comment by gerion0** [comment]
The same holds for the `custom_tgt` object that `fs.copyfile()` generates. Is it meaningful to _always_ create a dependency when a `full_path()` of a target or file is taken?

In concrete I exprienced this in code like this:
```meson
h = fs.copyfile(some_header, 'this.h')
my_include = ['-I', h.full_path()]
foo = custom_target(..., command: [custom_compiler] + my_include + ...)
```
In this case `foo` has no dependency to `h`.

**Comment by eli-schwartz** [comment]
This is really just a special case of https://mesonbuild.com/FAQ.html#how-do-i-tell-meson-that-my-sources-use-generated-headers

Consider a configure_file header that contains the path to prog1 and is `#include`d in prog2.c; it is not necessary to build and link prog1 before compiling prog2, but you do need prog1 to be built before you can *run* prog2, since the latter will try to `fork` + `exec` the former. However, if you configure the path to `#include "builddir/other-generated-header.h"` then one header does indeed depend on another header at *compile* time, and the standard solution to this is to list both as sources in any target that needs either one. Actually, you can make this easier with `declare_dependency(sources: [header1, header2])`.

There is no custom_target equivalent of declare_dependency(). It sounds like you want one.


Anyways: using the `.full_path()` method to cast an object to string will explicitly drop all dependency information, should dependency information exist, hence that is a moot point.

**Comment by gerion0** [comment]
> This is really just a special case of https://mesonbuild.com/FAQ.html#how-do-i-tell-meson-that-my-sources-use-generated-headers

You are right but unfortunately, I cannot use the standard Meson `library` and `exectable` functions and have to invoke the compiler by myself, since I want to compile the same C-Code with Clang to LLVM-IR-Code in different architectures (ARM, RISC-V, in the same project, and together with some other "normal" amd64 exectables, so Meson's cross compile features do not work here, beside of missing functionality to output LLVM-IR Code).

> There is no custom_target equivalent of declare_dependency(). It sounds like you want one.

I'm not quite sure what you mean with that. I could use `declare_dependency` for those objects. They contain all the necessary information and normally Meson automagically generates the correct include flags, linker flags and build dependencies of a given internal dependency. However, since I have no possibility to get the internal values of a dependency back to build my custom compiler call, I cannot use it (I already created #12543 for that).


> Anyways: using the `.full_path()` method to cast an object to string will explicitly drop all dependency information, should dependency information exist, hence that is a moot point.

Ok, I get your point. Using a dependency object but be able to use it in a `custom_target` would be a much cleaner way.

BTW, this
```
h = fs.copyfile(some_header, 'this.h')
my_include = ['-I', h.full_path()]
```
is just a workround for not yet being able to call `full_path()` on files directly. I saw that this is possible in the yet unreleased next Meson version.


> Consider a configure_file header that contains the path to prog1 and is `#include`d in prog2.c; it is not necessary to build and link prog1 before compiling prog2, but you do need prog1 to be built before you can _run_ prog2, since the latter will try to `fork` + `exec` the former. However, if you configure the path to `#include "builddir/other-generated-header.h"` then one header does indeed depend on another header at _compile_ time, and the standard solution to this is to list both as sources in any target that needs either one. Actually, you can make this easier with `declare_dependency(sources: [header1, header2])`.

Hmm, to summarize it for my initial example in this issue: The run target `cat` needs both targets at runtime: `sh` and `some_tgt`, but `sh` is perfectly buildable without the execution of `some_tgt`. But I cannot use `declare_dependancy` here to bound `sh` and `some_tgt` together, since I cannot get back the internal values in the `run_target`.

BTW: I discovered while researching for this issue. `fs.copyfile()` returns a `custom_tgt` which makes sense (however, one cannot specify some values like `build_always_stale` or `build_by_default` for it). `configure_file()`, however, returns a `file` object. Normally, `file` objects always describe already existing files in the source directory, while `configure_file` returns a generated file in the build directory. Wouldn't some form of a `tgt` object be better suited for its return value?

**Comment by gerion0** [comment]
> > > There is no custom_target equivalent of declare_dependency(). It sounds like you want one.
> > 
> > 
> > I'm not quite sure what you mean with that. I could use `declare_dependency` for those objects. They contain all the necessary information and normally Meson automagically generates the correct include flags, linker flags and build dependencies of a given internal dependency. However, since I have no possibility to get the internal values of a dependency back to build my custom compiler call, I cannot use it (I already created #12543 for that).
> 
> Basically something like `declare_custom_target_args(sh, depends: some_tgt)`. You'd define an array of `commands: ` compatible arguments that could have associated metadata attached to its use as an array passed to custom_target.

I was wondering what the return type of `declare_custom_target_args` would be. Has it a `full_path()` function, if the input was a file object (i.e. does it behave like a file but tracks the additional dependency)? And does it behave differently if a string was given. Can I add it to another string then?
`declare_custom_target_args` seems to be restricted to a single argument. I imagine that it could be quite cumbersome to use it with a list of arguments (although I'm not quite sure about it).
Wouldn't be the existing `declare_dependency()` better suited in the generic case (it is made exactly for coupling related things together, isn't it?) as long as one has a possibility to get the arguments back?

Maybe something like:
```
my_dep = declare_dependency(extra_files: sh, dependencies: some_tgt)
...
run_target('cat', command: [find_program('bash'), my_dep.extra_files()[0]], depends: my_dep)
```

You have to specify the extra dependency to `my_dep` here explicitly but in contrast to the original example there is not need to remember the dependency connection between `sh` and `some_tgt` anymore.


> > BTW: I discovered while researching for this issue. `fs.copyfile()` returns a `custom_tgt` which makes sense (however, one cannot specify some values like `build_always_stale` or `build_by_default` for it). `configure_file()`, however, returns a `file` object. Normally, `file` objects always describe already existing files in the source directory, while `configure_file` returns a generated file in the build directory. Wouldn't some form of a `tgt` object be better suited for its return value?
> 
> A file object describes existing files at the time of `meson setup` whereas tgt objects describe files which are created by ninja.
> 
> This is "usually" about whether the files are in the source vs build directory, but there are some files created in the build directory as part of "setup" and not by ninja. Primarily, configure_file outputs.

Ah, thanks for the explanation. I'm assuming that Meson generates the configure_file outputs at setup time to fulfil compiler dependencies at their first run (so without a dependency file)?  So let's take this example (of the manual):
```meson
conf_data = configuration_data()
# Set data
configure_file(input : 'projconfig.h.in',
  output : 'projconfig.h',
  configuration : conf_data)
configuration_inc = include_directories('.')
executable(..., include_directories : configuration_inc)
```
Meson does not know that the compiler expects `projconfig.h` in `.` so in pregenerates it before any Ninja call?

Wouldn't this fix it, if configure_file would be a target?
```meson
conf_data = configuration_data()
# Set data
conf_file = configure_file(input : 'projconfig.h.in',
  output : 'projconfig.h',
  configuration : conf_data)
conf_dep = declare_dependency(include_directores: '.', dependencies: conf_file)
executable(..., depends: conf_dep)

**Comment by eli-schwartz** [comment]
> Ah, thanks for the explanation. I'm assuming that Meson generates the configure_file outputs at setup time to fulfil compiler dependencies at their first run (so without a dependency file)? So let's take this example (of the manual):

Yes and no. Usually a `config.h` is widely used by most everything, and adding dependencies everywhere gets old quickly.

But there's also the fact that the input to a configure_file is state information from meson.build itself. While in theory this *could* be delayed until build time, that would require meson to write out a python script at setup time containing the configuration_data() values so that the python script could produce `config.h`. This is not usually valuable to do because the traditional use case of configure_file is typically for files that don't change often.

If you do want to configure files at build time instead of setup time, I'd recommend implementing something like this:
https://git.sr.ht/~eschwartz/pacman/tree/master/item/build-aux/edit-script.sh.in

That would then get configured with configure_file at setup time, and you'd use it as a custom_target() command to build other files at build time.
